src/anlz.rs

```diff
@@ -775,522 +531,315 @@ pub struct Phrase {
-            .cloned()
-            .map(WaveformPreviewColumn::from)
-            .collect();
-
-        Ok((
-            input,
-            Content::WaveformDetail {
-                len_entry_bytes,
-                len_entries,
-                unknown,
-                data,
-            },
-        ))
-    }
-
-    fn parse_waveform_color_preview<'a>(
-        input: &'a [u8],
-        _header: &Header,
-    ) -> IResult<&'a [u8], Self> {
-        let (input, len_entry_bytes) = nom::number::complete::be_u32(input)?;
-        // All waveform color preview entries should be 6 bytes long. If we see other values here,
-        // some reverse-engineering is needed.
-        if len_entry_bytes != 2 {
-            return Err(nom_input_error_with_kind(input, ErrorKind::LengthValue));
-        }
-
-        let (input, len_entries) = nom::number::complete::be_u32(input)?;
-        let entry_count = usize::try_from(len_entries)
-            .map_err(|_| nom_input_error_with_kind(input, ErrorKind::TooLarge))?;
-
-        let (input, unknown) = nom::number::complete::be_u32(input)?;
-        let (input, data) =
-            nom::multi::count(WaveformColorPreviewColumn::parse, entry_count)(input)?;
-
-        Ok((
-            input,
-            Content::WaveformColorPreview {
-                len_entry_bytes,
-                len_entries,
-                unknown,
-                data,
-            },
-        ))
-    }
-
-    fn parse_waveform_color_detail<'a>(
-        input: &'a [u8],
-        _header: &Header,
-    ) -> IResult<&'a [u8], Self> {
-        let (input, len_entry_bytes) = nom::number::complete::be_u32(input)?;
-        // All waveform color detail entries should be 2 bytes long. If we see other values here,
-        // some reverse-engineering is needed.
-        if len_entry_bytes != 2 {
-            return Err(nom_input_error_with_kind(input, ErrorKind::LengthValue));
-        }
-
-        let (input, len_entries) = nom::number::complete::be_u32(input)?;
-        let entry_count = usize::try_from(len_entries)
-            .map_err(|_| nom_input_error_with_kind(input, ErrorKind::TooLarge))?;
-
-        let (input, unknown) = nom::number::complete::be_u32(input)?;
-        let (input, data) =
-            nom::multi::count(WaveformColorDetailColumn::parse, entry_count)(input)?;
-
-        Ok((
-            input,
-            Content::WaveformColorDetail {
-                len_entry_bytes,
-                len_entries,
-                unknown,
-                data,
-            },
-        ))
-    }
+/// All beats in the track.
+#[derive(Debug, PartialEq)]
+#[binrw]
+pub struct BeatGrid {
+    /// Unknown field.
+    unknown1: u32,
+    /// Unknown field.
+    ///
+    /// According to [@flesniak](https://github.com/flesniak), this is always `00800000`.
+    unknown2: u32,
+    /// Number of beats in this beatgrid.
+    len_beats: u32,
+    /// Beats in this beatgrid.
+    #[br(count = len_beats)]
+    pub beats: Vec<Beat>,
```

Comment:
note that exposing the `Vec` publicly mutably means that the `beats.len() == len_beats` invariant can be violated. `binrw` currently does not have a way to fix these up easily. see the issue where this is discussed in the context of `binrw::FilePtr`: https://github.com/jam1garner/binrw/issues/4

Reply:
> note that exposing the `Vec` publicly mutably means that the `beats.len() == len_beats` invariant can be violated.

Isn't that the case anyway even if the field was not `pub`? I could still crate an instance of the struct with mismatching len_beats and beats values by hand. The fix would be to make `len_beats` a temp field that isn't actually part of the struct and is only parsed for reading and automatically calculated from `beats` when writing.

Reply:
Not sure, as mentioned in the FilePtr issue, a second pass could probably help with most of these issues. Another one could be to pass a reference to the field in the types args, but that is not supported either currently. Though both are on the v0.9.0 roadmap https://github.com/users/jam1garner/projects/1 

> Isn't that the case anyway even if the field was not pub? I could still crate an instance of the struct with mismatching len_beats and beats values by hand.

well if the type is public then yes, for this reason, I opted for an encapsulation approach in the DeviceSQLString implementation where the string is constructed correctly once and then should never be manipulated again, at least thats the plan, there are still some kinks to work out.

Reply:
Should be fixed with https://github.com/Holzhaus/rekordcrate/pull/51.